File: kotlin/src/main/kotlin/foundation/zoobc/zbc/Ed25519.kt
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci
package foundation.zoobc.zbc

import java.math.BigInteger
import java.security.MessageDigest

/**
 * Ed25519 (RFC 8032) on BigInteger: public key from a 32-byte seed, detached signatures,
 * verification. The scalar multiplication takes the same path for every scalar at the Kotlin
 * level; BigInteger is not constant time, so keep seeds on machines you trust.
 */
object Ed25519 {
    private val P = BigInteger.TWO.pow(255) - BigInteger.valueOf(19)
    private val L = BigInteger.TWO.pow(252) + BigInteger("27742317777372353535851937790883648493")
    private val D = BigInteger.valueOf(-121665).multiply(BigInteger.valueOf(121666).modInverse(P)).mod(P)
    private val I = BigInteger.TWO.modPow((P - BigInteger.ONE).divide(BigInteger.valueOf(4)), P)

    private class Pt(val x: BigInteger, val y: BigInteger, val z: BigInteger, val t: BigInteger)

    private fun recoverX(y: BigInteger, sign: Int): BigInteger {
        val y2 = y.multiply(y).mod(P)
        val u = (y2 - BigInteger.ONE).mod(P)
        val v = (D.multiply(y2) + BigInteger.ONE).mod(P)
        var x = u.multiply(v.modInverse(P)).modPow((P + BigInteger.valueOf(3)).divide(BigInteger.valueOf(8)), P)
        if (v.multiply(x).multiply(x).subtract(u).mod(P) != BigInteger.ZERO) x = x.multiply(I).mod(P)
        if (v.multiply(x).multiply(x).subtract(u).mod(P) != BigInteger.ZERO) throw IllegalArgumentException("not a point on the curve")
        if (x.testBit(0) != (sign == 1)) x = P - x
        return x
    }

    private val GY = BigInteger.valueOf(4).multiply(BigInteger.valueOf(5).modInverse(P)).mod(P)
    private val GX = recoverX(GY, 0)
    private val G = Pt(GX, GY, BigInteger.ONE, GX.multiply(GY).mod(P))
    private val ZERO = Pt(BigInteger.ZERO, BigInteger.ONE, BigInteger.ONE, BigInteger.ZERO)

    private fun add(p: Pt, q: Pt): Pt {
        val a = (p.y - p.x).multiply(q.y - q.x).mod(P)
        val b = (p.y + p.x).multiply(q.y + q.x).mod(P)
        val c = BigInteger.TWO.multiply(p.t).multiply(q.t).multiply(D).mod(P)
        val d = BigInteger.TWO.multiply(p.z).multiply(q.z).mod(P)
        val e = b - a; val f = d - c; val g = d + c; val h = b + a
        return Pt(e.multiply(f).mod(P), g.multiply(h).mod(P), f.multiply(g).mod(P), e.multiply(h).mod(P))
    }

    private fun mul(p: Pt, s: BigInteger): Pt {
        var r = ZERO; var q = p
        for (i in 0 until 256) {
            val sum = add(r, q)
            r = if (s.testBit(i)) sum else r
            q = add(q, q)
        }
        return r
    }

    private fun encode(p: Pt): ByteArray {
        val zi = p.z.modInverse(P)
        val x = p.x.multiply(zi).mod(P); val y = p.y.multiply(zi).mod(P)
        val v = if (x.testBit(0)) y.setBit(255) else y
        return toLe(v, 32)
    }

    private fun decode(b: ByteArray): Pt {
        require(b.size == 32) { "point must be 32 bytes" }
        val raw = fromLe(b)
        val sign = if (raw.testBit(255)) 1 else 0
        val y = raw.clearBit(255)
        if (y >= P) throw IllegalArgumentException("non-canonical point")
        val x = recoverX(y, sign)
        if (x == BigInteger.ZERO && sign == 1) throw IllegalArgumentException("non-canonical point")
        return Pt(x, y, BigInteger.ONE, x.multiply(y).mod(P))
    }

    private fun fromLe(b: ByteArray): BigInteger = BigInteger(1, b.reversedArray())
    private fun toLe(v: BigInteger, n: Int): ByteArray {
        val be = v.toByteArray()
        val out = ByteArray(n)
        var j = 0
        for (i in be.indices.reversed()) { if (j < n) out[j++] = be[i] }
        return out
    }
    private fun sha512(vararg parts: ByteArray): ByteArray = MessageDigest.getInstance("SHA-512").also { md -> parts.forEach { md.update(it) } }.digest()
    private fun clamp(h: ByteArray): BigInteger {
        val k = h.copyOf(32)
        k[0] = (k[0].toInt() and 248).toByte(); k[31] = (k[31].toInt() and 127).toByte(); k[31] = (k[31].toInt() or 64).toByte()
        return fromLe(k)
    }

    /** The 32-byte public key of a 32-byte seed. */
    fun publicKey(seed: ByteArray): ByteArray {
        require(seed.size == 32) { "seed must be 32 bytes" }
        return encode(mul(G, clamp(sha512(seed))))
    }

    /** Detached 64-byte signature of [message] with the 32-byte seed. */
    fun sign(message: ByteArray, seed: ByteArray): ByteArray {
        require(seed.size == 32) { "seed must be 32 bytes" }
        val h = sha512(seed)
        val a = clamp(h)
        val prefix = h.copyOfRange(32, 64)
        val pub = encode(mul(G, a))
        val r = fromLe(sha512(prefix, message)).mod(L)
        val bigR = encode(mul(G, r))
        val k = fromLe(sha512(bigR, pub, message)).mod(L)
        return bigR + toLe((r + k.multiply(a)).mod(L), 32)
    }

    /** True when [signature] is a valid signature of [message] by [pub]. Never throws. */
    fun verify(message: ByteArray, signature: ByteArray, pub: ByteArray): Boolean = try {
        if (signature.size != 64 || pub.size != 32) false else {
            val a = decode(pub)
            val bigR = decode(signature.copyOf(32))
            val s = fromLe(signature.copyOfRange(32, 64))
            if (s >= L) false else {
                val k = fromLe(sha512(signature.copyOf(32), pub, message)).mod(L)
                encode(mul(G, s)).contentEquals(encode(add(bigR, mul(a, k))))
            }
        }
    } catch (e: Exception) { false }
}

File: py/zbc/ed25519.py
# SPDX-License-Identifier: MIT
# Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci
"""Ed25519 (RFC 8032): public key from a 32-byte seed, detached signatures, verification.

Pure Python integers. The scalar multiplication takes the same path for every scalar at the
Python level; Python's big integers are not constant time, so keep seeds on machines you trust,
as with any pure-Python signer.
"""
import hashlib

P = 2 ** 255 - 19
L = 2 ** 252 + 27742317777372353535851937790883648493
D = (-121665 * pow(121666, P - 2, P)) % P
I = pow(2, (P - 1) // 4, P)


def _recover_x(y: int, sign: int) -> int:
    y2 = y * y % P
    u, v = (y2 - 1) % P, (D * y2 + 1) % P
    x = pow(u * pow(v, P - 2, P), (P + 3) // 8, P)
    if (v * x * x - u) % P != 0:
        x = x * I % P
    if (v * x * x - u) % P != 0:
        raise ValueError("not a point on the curve")
    if (x & 1) != sign:
        x = P - x
    return x


_GY = 4 * pow(5, P - 2, P) % P
_GX = _recover_x(_GY, 0)
_G = (_GX, _GY, 1, _GX * _GY % P)
_ZERO = (0, 1, 1, 0)


def _add(p, q):
    x1, y1, z1, t1 = p
    x2, y2, z2, t2 = q
    a = (y1 - x1) * (y2 - x2) % P
    b = (y1 + x1) * (y2 + x2) % P
    c = 2 * t1 * t2 * D % P
    d = 2 * z1 * z2 % P
    e, f, g, h = b - a, d - c, d + c, b + a
    return (e * f % P, g * h % P, f * g % P, e * h % P)


def _mul(p, s: int):
    r, q = _ZERO, p
    for i in range(256):
        total = _add(r, q)
        r = total if (s >> i) & 1 else r
        q = _add(q, q)
    return r


def _encode(p) -> bytes:
    zi = pow(p[2], P - 2, P)
    x, y = p[0] * zi % P, p[1] * zi % P
    return (y | ((x & 1) << 255)).to_bytes(32, "little")


def _decode(b: bytes):
    if len(b) != 32:
        raise ValueError("point must be 32 bytes")
    y = int.from_bytes(b, "little")
    sign = y >> 255
    y &= (1 << 255) - 1
    if y >= P:
        raise ValueError("non-canonical point")
    x = _recover_x(y, sign)
    if x == 0 and sign == 1:
        raise ValueError("non-canonical point")
    return (x, y, 1, x * y % P)


def _clamp(h: bytes) -> int:
    k = bytearray(h[:32])
    k[0] &= 248
    k[31] &= 127
    k[31] |= 64
    return int.from_bytes(k, "little")


def public_key(seed: bytes) -> bytes:
    """The 32-byte public key of a 32-byte seed."""
    if len(seed) != 32:
        raise ValueError("seed must be 32 bytes")
    return _encode(_mul(_G, _clamp(hashlib.sha512(seed).digest())))


def sign(message: bytes, seed: bytes) -> bytes:
    """Detached 64-byte signature of `message` with the 32-byte seed."""
    if len(seed) != 32:
        raise ValueError("seed must be 32 bytes")
    h = hashlib.sha512(seed).digest()
    a, prefix = _clamp(h), h[32:]
    pub = _encode(_mul(_G, a))
    r = int.from_bytes(hashlib.sha512(prefix + message).digest(), "little") % L
    big_r = _encode(_mul(_G, r))
    k = int.from_bytes(hashlib.sha512(big_r + pub + message).digest(), "little") % L
    return big_r + ((r + k * a) % L).to_bytes(32, "little")


def verify(message: bytes, signature: bytes, pub: bytes) -> bool:
    """True when `signature` is a valid signature of `message` by `pub`. Never raises."""
    try:
        if len(signature) != 64 or len(pub) != 32:
            return False
        a = _decode(pub)
        big_r = _decode(signature[:32])
        s = int.from_bytes(signature[32:], "little")
        if s >= L:
            return False
        k = int.from_bytes(hashlib.sha512(signature[:32] + pub + message).digest(), "little") % L
        return _encode(_mul(_G, s)) == _encode(_add(big_r, _mul(a, k)))
    except (ValueError, ArithmeticError):
        return False
